### PR TITLE
Create Archunit rule to detect if JUnit annotations are mixed within tests [HZ-3048]

### PR DESCRIPTION
**Pull request by orcunc**
Jira : https://hazelcast.atlassian.net/browse/HZ-3048

Breaking changes (list specific methods/types/messages):
* API
* client protocol format
* serialized form
* snapshot format

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+public class MixTestAnnotationsCondition extends ArchCondition<JavaClass> {
+    public MixTestAnnotationsCondition() {
+        super("Do not mix Junit4 and Junit5 annotations");
+    }
+
+    @Override
+    public void check(JavaClass item, ConditionEvents events) {
+        boolean hasJUnit4Annotation = item.getMethods().stream()
+                .anyMatch(method -> method.isAnnotatedWith(org.junit.Test.class)
+                                    || method.isAnnotatedWith(org.junit.Before.class)
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java]: rev : 200cf7e61351df1a647fbed099f1863075e39869

File: hazelcast-archunit-rules/src/test/java/com/hazelcast/test/archunit/ArchUnitRulesTest.java [hunk 2]
```diff
@@ -47,4 +47,27 @@ public void should_NOT_fail_with_non_compliant_class() {
...
     }
+
+    @Test
+    public void should_fail_with_mixed_annotation_test_class() {
+        JavaClasses classes = new ClassFileImporter()
+                .withImportOption(ONLY_INCLUDE_TESTS)
+                .importPackages("com.example.broken");
+        assertThat(classes).isNotEmpty();
+
+
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-archunit-rules/src/test/java/com/hazelcast/test/archunit/ArchUnitRulesTest.java]: rev : 200cf7e61351df1a647fbed099f1863075e39869

**Comment by orcunc** [comment]
run-lab-run

File: hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.lang.annotation.Annotation;
+import java.util.Set;
+
+public class MixTestAnnotationsCondition extends ArchCondition<JavaClass> {
+    private final Set<Class<? extends Annotation>> junit4AnnotationClasses = Set.of(Test.class, Before.class,
```
Comment (ldziedziul) [inline hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java:33]: maybe add it as static fields?

File: hazelcast/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.hazelcast.test.archunit.ArchUnitRules;
+import com.hazelcast.test.archunit.ArchUnitTestSupport;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import org.junit.Test;
+
+import static com.hazelcast.test.archunit.ModuleImportOptions.onlyCurrentModuleTests;
+
+public class NoMixedJUnitAnnotationsInOurTestSourcesTest extends ArchUnitTestSupport {
```
Comment (ldziedziul) [inline hazelcast/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java:27]: What about other maven modules?

Comment (orcunc) [reply hazelcast/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java:27]: The other modules for example hazelcast-sql, hazelcast-tpc-engine 
do not have dependency for archunit in their pom.xml files. None of them have archunit rule. I am not sure why

File: hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java (hunk 3 above)
Comment (orcunc) [reply hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java]: rev : c2c4477d9c54cbd9205f153917bc9324d9724db4

File: hazelcast/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java (hunk 4 above)
Comment (orcunc) [reply hazelcast/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java:27]: rev : 7dbaaa4de9a8696be18007b93ebbc488fb133d3e

**Comment by orcunc** [comment]
run-lab-run

**Comment by orcunc** [comment]
run-sonar

**Comment by orcunc** [comment]
run-sonar

**Comment by olukas** [comment]
run-sonar

**Review by ldziedziul — APPROVED** [review]

**Review by TomaszGaweda — APPROVED** [review]

**Pull request merged**